File: deck.go
// Modelling a deck of 52 cards
package main

func generateDeck() []card {
	suits := [4]string{"Hearts", "Spades", "Clubs", "Diamonds"}
	currentSuit := 0

	var deck []card

	for i := 0; i < 4; i++ {
		for j := 2; j < 15; j++ {
			deck = append(deck, card{rank: j, suit: suits[currentSuit]})
		}
		currentSuit++
	}

	return deck
}

File: lib/deck.rb
class Deck
  def initialize
    @deck = []
    suits = ["Hearts", "Spades", "Clubs", "Diamonds"]
    currentSuit = 0
    
    # Initialise a full deck
    while currentSuit < 4
      i = 1
      while i < 14
        c = Card.new(i, suits[currentSuit])
        @deck << c
        i += 1
      end
      currentSuit += 1
    end
  end

  def deck
    @deck
  end

  def shuffle
    @deck.shuffle!
  end
end